File: 2024/day06-2.ts
type Coord = {
  x: number;
  y: number;
};
type Direction = "^" | "v" | "<" | ">";
type Guard = Coord & { direction: Direction };

function solution(content: string) {
  const rows = content.split("\n").length;
  const startPosition = parseStartPosition(rows, content);
  const obstacles = parseObstacles(rows, content);
  const limits = parseLimits(rows, content);
  const positions: Coord[] = track(obstacles, limits, startPosition);

  return positions
    .map((position) => play([...obstacles, position], limits, startPosition))
    .map((item) => (item ? 1 : 0))
    .reduce<number>((acc, cur) => acc + cur, 0);
}

function play(obstacles: Coord[], limits: Coord, start: Guard) {
  let current = start;
  const positions: Guard[] = [start];
  while (true) {
    const newCurrent = move(current);
    if (isOutside(limits, newCurrent)) {
      return false;
    }
    if (
      positions.find(
        (pos) =>
          pos.x === newCurrent.x &&
          pos.y === newCurrent.y &&
          pos.direction === newCurrent.direction
      )
    ) {
      return true;
    }
    if (isObstacle(obstacles, newCurrent)) {
      current = rotate(current);
    } else {
      current = newCurrent;
      positions.push(current);
    }
  }
}

function track(obstacles: Coord[], limits: Coord, start: Guard) {
  let current = start;
  const positions: Guard[] = [start];
  while (true) {
    const newCurrent = move(current);
    if (isOutside(limits, newCurrent)) {
      break;
    } else if (isObstacle(obstacles, newCurrent)) {
      current = rotate(current);
    } else {
      current = newCurrent;
      positions.push(current);
    }
  }
  return Array.from(
    new Set(
      positions
        .filter(
          (item, index, array) =>
            !array
              .slice(0, index)
              .find((itm) => itm.x === item.x && itm.y === item.y)
        )
        .map(({ x, y }) => ({ x, y }))
    )
  );
}

function rotate(current: Guard): Guard {
  const { x, y, direction } = current;
  if (direction === "^") {
    return { x, y, direction: ">" };
  }
  if (direction === "v") {
    return { x, y, direction: "<" };
  }
  if (direction === "<") {
    return { x, y, direction: "^" };
  }
  if (direction === ">") {
    return { x, y, direction: "v" };
  }
  throw new Error(`Invalid direction: ${direction}`);
}

function move(current: Guard): Guard {
  const { x, y, direction } = current;
  if (direction === "^") {
    return { x, y: y - 1, direction };
  }
  if (direction === "v") {
    return { x, y: y + 1, direction };
  }
  if (direction === "<") {
    return { x: x - 1, y, direction };
  }
  if (direction === ">") {
    return { x: x + 1, y, direction };
  }
  throw new Error(`Invalid direction: ${direction}`);
}

function isObstacle(obstacles: Coord[], current: Guard) {
  const { x, y } = current;
  return !!obstacles.find((item) => item.x === x && item.y === y);
}

function isOutside(limits: Coord, current: Guard) {
  const { x: right, y: bottom } = limits;
  const { x, y } = current;
  return x < 0 || y < 0 || x > right || y > bottom;
}

function parseLimits(rows: number, content: string): Coord {
  return indexToCoord(
    rows,
    content.split("").filter((item) => item !== "\n").length - 1
  );
}

function parseObstacles(rows: number, content: string): Coord[] {
  return content
    .split("")
    .filter((item) => item !== "\n")
    .map((item, index) => {
      if (item === "#") return indexToCoord(rows, index);
      return null;
    })
    .filter((item) => item !== null);
}

function parseStartPosition(rows: number, content: string): Guard {
  const direction = content
    .split("")
    .find((item) => "^v<>".includes(item)) as Direction;
  const index = content
    .split("")
    .filter((item) => item !== "\n")
    .findIndex((item) => "^v<>".includes(item));
  const { x, y } = indexToCoord(rows, index);
  return {
    x,
    y,
    direction,
  };
}

function indexToCoord(rows: number, index: number): Coord {
  return {
    x: index % rows,
    y: Math.floor(index / rows),
  };
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2024/day06-2.hs
import Data.List (find, findIndex)
import Data.Maybe (catMaybes, fromJust, isJust, mapMaybe)
import Data.Set (fromList, toList)

type Coord = (Int, Int)

type Guard = (Int, Int, Char)

solution :: String -> Int
solution content = (length . filter (\p -> play (p : obstacles) limits startPosition)) positions
  where
    rows = (length . lines) content
    startPosition = parseStartPosition rows content
    obstacles = parseObstacles rows content
    limits = parseLimits rows content
    positions = (toList . fromList . map (\(x, y, _) -> (x, y))) (track obstacles limits startPosition)

play :: [Coord] -> Coord -> Guard -> Bool
play obstacles limits current = helper [current] obstacles limits current
  where
    helper :: [Guard] -> [Coord] -> Coord -> Guard -> Bool
    helper positions obstacles limits current
      | isOutside limits newCurrent = False
      | newCurrent `elem` positions = True
      | isObstacle obstacles newCurrent = helper positions obstacles limits (rotate current)
      | otherwise = helper (newCurrent : positions) obstacles limits newCurrent
      where
        newCurrent = move current

track :: [Coord] -> Coord -> Guard -> [Guard]
track obstacles limits current = helper [current] obstacles limits current
  where
    helper :: [Guard] -> [Coord] -> Coord -> Guard -> [Guard]
    helper positions obstacles limits current
      | isOutside limits newCurrent = positions
      | isObstacle obstacles newCurrent = helper positions obstacles limits (rotate current)
      | otherwise = helper (newCurrent : positions) obstacles limits newCurrent
      where
        newCurrent = move current

rotate :: Guard -> Guard
rotate (x, y, '^') = (x, y, '>')
rotate (x, y, 'v') = (x, y, '<')
rotate (x, y, '<') = (x, y, '^')
rotate (x, y, '>') = (x, y, 'v')

move :: Guard -> Guard
move (x, y, '^') = (x, y - 1, '^')
move (x, y, 'v') = (x, y + 1, 'v')
move (x, y, '<') = (x - 1, y, '<')
move (x, y, '>') = (x + 1, y, '>')

isObstacle :: [Coord] -> Guard -> Bool
isObstacle obstacles (x, y, _) = isJust (find (\(ox, oy) -> x == ox && y == oy) obstacles)

isOutside :: Coord -> Guard -> Bool
isOutside (right, bottom) (x, y, _) = x < 0 || y < 0 || x > right || y > bottom

parseLimits :: Int -> String -> Coord
parseLimits rows content = indexToCoord rows (length (filter (/= '\n') content) - 1)

parseObstacles :: Int -> String -> [Coord]
parseObstacles rows content =
  catMaybes
    ( zipWith
        ( \index item ->
            if' (Just (indexToCoord rows index)) Nothing (item == '#')
        )
        [0 ..]
        (filter (/= '\n') content)
    )

parseStartPosition :: Int -> String -> Guard
parseStartPosition rows content = (x, y, d)
  where
    index = fromJust (findIndex (`elem` "^v<>") (filter (/= '\n') content))
    (x, y) = indexToCoord rows index
    d = fromJust (find (`elem` "^v<>") content)

indexToCoord :: Int -> Int -> Coord
indexToCoord rows index = (index `mod` rows, floor (fromIntegral index / fromIntegral rows))

if' :: a -> a -> Bool -> a
if' a _ True = a
if' _ a False = a

main :: IO ()
main = do readFile "day06.dat" >>= print . solution
